### PR TITLE
Add support for float and double types.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
* Adds Encode and Decode support for float and double.

* Adds support for float/double in various code generation type
  helpers (for chip-tool, java, darwin).

* Adds float and double members to the SimpleStruct test struct.

* Adds a command to send and receive a SimpleStruct, and a yaml test
  that does that.

* Fixes chip-tool to handle BitFlags value checking correctly (since
  SimpleStruct has a BitFlags member).

* Removes some code in ClustersHelper.js that failed for float/double
  types and whose output was completely unused.

* Fixes templates for old-style commands and java that completely blew
  up on structs in a response command field to at least generate code
  successfully, though that code wouldn't work right if it ever ran.

#### Problem
Codegen fails if someone uses the "single" or "double" type in an XML file.

#### Change overview
See above.

#### Testing
New yaml test, plus some code inspection of the generated code.

### CODE REVIEW

**Comment by bzbarsky-apple** [comment]
Note: This does NOT add support for float/double typed attributes yet.  This PR had gotten way too big as it was....  That will need to be a followup.

File: examples/chip-tool/templates/commands.zapt [hunk 1]
```diff
@@ -46,6 +46,13 @@ CHIP_ERROR LogValue(const char * label, size_t indent, X value)
     return CHIP_NO_ERROR;
 }
 
+template <typename X, typename std::enable_if_t<std::is_floating_point<X>::value && !std::is_same<std::remove_cv_t<std::remove_reference_t<X>>, bool>::value, int> = 0>
```
Comment (Damian-Nordic) [inline examples/chip-tool/templates/commands.zapt:49]: If think if `is_floating_point<X>` then we can be sure that `X` is not bool, so the further part of the condition is not needed.

File: src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt [hunk 2]
```diff
@@ -8,7 +8,15 @@ wasHandled = emberAf{{asUpperCamelCase parent.name}}Cluster{{asUpperCamelCase na
 {{#if (zcl_command_arguments_count this.id)}}
 expectArgumentCount = {{ zcl_command_arguments_count this.id }};
 {{#zcl_command_arguments}}
-{{asUnderlyingZclType type}} {{asSymbol label}};
+  {{#if isArray}}
+    {{asUnderlyingZclType type}} {{asSymbol label}};
+  {{else}}
```
Comment (Damian-Nordic) [inline src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt:13]: `{{else if...}}`?

**Review by Damian-Nordic — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

File: examples/chip-tool/templates/commands.zapt (hunk 1 above)
Comment (bzbarsky-apple) [reply examples/chip-tool/templates/commands.zapt:49]: Good catch, fixed.

File: src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt:13]: No, because you can't do `else if_is_struct type`, sadly.

Comment (Damian-Nordic) [reply src/app/zap-templates/partials/im_command_handler_cluster_commands.zapt:13]: Ah, I missed the underscore.

**Comment by andy31415** [comment]
fast track: change created and approved by domain owners

**Pull request merged**